src/Lorentz/Contracts/BaseDAO/Types.hs

```diff
@@ -40,73 +155,209 @@ instance TypeHasDoc MigrationStatus where
   | Accept_ownership ()
   | Migrate MigrateParam
   | Confirm_migration ()
+  | Propose (ProposeParams proposalMetadata)
+  | Proposal_metadata (View ProposalKey (Proposal proposalMetadata))
+  | Vote [VoteParam]
+  -- Admin
+  | Set_voting_period VotingPeriod
+  | Set_quorum_threshold QuorumThreshold
+  | Flush ()
   deriving stock (Generic, Show)
-  deriving anyclass (IsoValue)
 
-instance ParameterHasEntrypoints Parameter where
-  type ParameterEntrypointsDerivation Parameter = EpdDelegate
+instance (HasAnnotation pm, NiceParameter pm) => ParameterHasEntrypoints (Parameter pm) where
+  type ParameterEntrypointsDerivation (Parameter pm) = EpdDelegate
+
+deriving anyclass instance (WellTypedIsoValue pm) => IsoValue (Parameter pm)
 
 type TransferOwnershipParam = ("newOwner" :! Address)
 type MigrateParam = ("newAddress" :! Address)
 
-emptyStorage :: Storage
+-- | Voting period in seconds
+type VotingPeriod = Natural
+
+-- | QuorumThreshold that a proposal need to meet
+-- A proposal will be rejected if the quorum_threshold is not met,
+-- regardless of upvotes > downvotes
+-- A proposal will be accepted only if the
+-- (quorum_threshold >= upvote + downvote) && (upvote > downvote)
+type QuorumThreshold = Natural
+
+emptyStorage :: Storage pm
 emptyStorage = Storage
   { sLedger =  BigMap $ Map.empty
   , sOperators = BigMap $ Map.empty
-  -- TODO: make it configurable
   , sTokenAddress = genesisAddress
   , sPendingOwner = Nothing
   , sAdmin = genesisAddress
   , sMigrationStatus = NotInMigration
+  , sVotingPeriod = 60 * 60 * 24 * 7 -- 7 days
+  , sQuorumThreshold = 4
+    -- ^ any proposals that have less that 4 votes will be rejected
+    -- regardless of upvotes
+
+  , sProposals = BigMap $ Map.empty
+  , sProposalKeyListSortByDate = []
   }
 
 mkStorage
   :: Address
   -> Map (Address, FA2.TokenId) Natural
   -> Operators
-  -> Storage
+  -> Storage pm
 mkStorage admin balances operators = Storage
   { sLedger = BigMap balances
   , sOperators = operators
   , sTokenAddress = genesisAddress
   , sPendingOwner = Nothing
   , sAdmin = admin
   , sMigrationStatus = NotInMigration
+  , sVotingPeriod = 60 * 60 * 24 * 7 -- days
+  , sQuorumThreshold = 4
+
+  , sProposals = BigMap $ Map.empty
+  , sProposalKeyListSortByDate = []
+  }
+
+------------------------------------------------------------------------
+-- Proposal
+------------------------------------------------------------------------
+
+type ProposalKey = ByteString
+
+-- | Proposal type which will be stored in 'Storage' `sProposals`
+-- `pVoters` is needed due to we need to keep track of voters to be able to
+-- unfreeze their tokens.
+data Proposal proposalMetadata = Proposal
+  { pUpvotes :: Natural
+  , pDownvotes :: Natural
+  , pStartDate :: Timestamp
+
+  , pMetadata :: proposalMetadata
+
+  , pProposer :: Address
+  , pProposerFrozenToken :: Natural
+
+  , pVoters :: [(Address, Natural)] -- This can be redundant with "Upvote/Downvote"
```

Comment:
> -- This can be redundant with "Upvote/Downvote"

What do you mean? AFAIU it's not redundant because we need to unfreeze tokens.

Reply:
That comment was outdated, I removed this.